pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -175,3 +195,146 @@ func (t *Terminator) podDeleteTimeWithGracePeriod(nodeGracePeriodExpirationTime
 	deleteTime := nodeGracePeriodExpirationTime.Add(time.Duration(*pod.Spec.TerminationGracePeriodSeconds) * time.Second * -1)
 	return &deleteTime
 }
+
+func (t *Terminator) GetDeploymentFromPod(ctx context.Context, pod *corev1.Pod) (*appsv1.Deployment, error) {
+	rs, err := t.getOwnerReplicaSet(ctx, pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ReplicaSet from Pod: %w", err)
+	}
+	if rs == nil {
+		return nil, nil
+	}
+
+	deployment, err := t.getOwnerDeployment(ctx, rs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Deployment from ReplicaSet: %w", err)
+	}
+	return deployment, nil
+
+}
+
+func (t *Terminator) getOwnerReplicaSet(ctx context.Context, pod *corev1.Pod) (*appsv1.ReplicaSet, error) {
+	for _, ownerRef := range pod.GetOwnerReferences() {
+		if ownerRef.Controller != nil && ownerRef.Kind == "ReplicaSet" {
+			rs := &appsv1.ReplicaSet{}
+			if err := t.kubeClient.Get(ctx, client.ObjectKey{Name: ownerRef.Name, Namespace: pod.Namespace}, rs); err != nil {
+				return nil, fmt.Errorf("get ReplicaSet: %w", err)
+			}
+			return rs, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (t *Terminator) getOwnerDeployment(ctx context.Context, rs *appsv1.ReplicaSet) (*appsv1.Deployment, error) {
+	for _, ownerRef := range rs.GetOwnerReferences() {
+		if ownerRef.Controller != nil && ownerRef.Kind == "Deployment" {
+			deployment := &appsv1.Deployment{}
+			if err := t.kubeClient.Get(ctx, client.ObjectKey{Name: ownerRef.Name, Namespace: rs.Namespace}, deployment); err != nil {
+				return nil, fmt.Errorf("get Deployment: %w", err)
+			}
+			return deployment, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (t *Terminator) RestartDeployments(ctx context.Context, deployments []*appsv1.Deployment, nodeName string) error {
+	var updateErrors []error
+
+	for _, deployment := range deployments {
+		if deployment.Spec.Template.Annotations == nil {
+			deployment.Spec.Template.Annotations = make(map[string]string)
+		}
+		restartedNode, exists := deployment.Spec.Template.Annotations["kubectl.kubernetes.io/restartedNode"]
```

Comment:
Why did we choose the name `kubectl.kubernetes.io/restartedNode` instead of something like x.karpenter.sh/xxx?

Reply:
It is OK to use "karpenter.sh/restartedNode". I refer to the restart method of k8s. It is more reasonable to change it to karpenter style here. Can you help review and see which parts are unreasonable and need to be modified? I will modify it together

Reply:
Sure, I will take a look asap.